Moved SpecialTokens assignment after the modification to avoid "Collection Modified" error

shaltielshmid opened the pull request:
We are excited to review your PR.

So we can do the best job, please check:

- [v] There's a descriptive title that will make sense to other developers some time from now. 
- [v] There's associated issues. All PR's should have issue(s) associated - unless a trivial self-evident change such as fixing a typo. You can use the format `Fixes #nnnn` in your description to cause GitHub to automatically close the issue(s) when your PR is merged.
- [v] Your change description explains what the change does, why you chose your approach, and anything else that reviewers should know.
- [ ] You have included any necessary tests in the same PR.

---

Hey all! I was working with BertTokenizer, and noticed that when I specified "BasicTokenization" and "Lowercase" then I was getting the "Collection Modified" error, since the updated dictionary is iterated over and updated in the same loop.
Solution: Just assign the dictionary after. 

I didn't include an issue/tests since this was just a simple typo fix. I'm happy to expand further if needed.

shaltielshmid commented:
I haven't tested this, just a theory - but a fear I have with the lowercase is that even though we identify the special tokens and keep them as a separate unit, because they are converted to lowercase they aren't going to be identified at the vocabulary lookup stage. 
I think we need to add it to the vocab as well, as is done in the AddSpecialToken function. @michaelgsharp what do you think?

tarekgh commented:
@shaltielshmid Thank you for catching that! Could you please add a test using the same code that reproduced the issue?

tarekgh commented:
I marked this for 4.0 as will be good to service this fix.

CC @ericstj

shaltielshmid commented:
@tarekgh Thank you for the response! 

Looking into it further, there seems to be a bigger question here. 

In the vocabulary, the special tokens are stored in uppercase (e.g., `[CLS]`, `[MASK]`, etc.). 

When the user does not specify the SpecialTokens, then the special tokens dictionary is created in the code. When this is done, the tokens are normalized before adding them to the SpecialTokens dictionary, and they are *also* appended to the vocabulary in their lowercase form:

https://github.com/dotnet/machinelearning/blob/81122c4c48d84fe2f49afd430a2c8fc214311baa/src/Microsoft.ML.Tokenizers/Model/BertTokenizer.cs#L715-L726

This solves the issue of identifying the lowercased special tokens when tokenizing the text, but also brings up a few questions:

1] There is a hidden assumption here that the SpecialTokens dictionary keys will be lowercased. 

2] What if the lowercased special token already existed in the vocabulary, as a regular token? Of course, this is not a very likely scenario, but worth considering. 

3] In the `WordPieceTokenizer.cs` file, a reversed dictionary is created (`SpecialTokensReverse`) - does it make sense that this dictionary value points to the lowercased token? 
- It seems like the usage of this dictionary is just to check if an ID is a special token, so to solve this we can just replace this variable with a HashSet of the IDs. 

This lead to an issue with the way the SpecialTokens are handled when specified explicitly:

If the SpecialTokens keys were passed in uppercase (as would be expected), then the code creates a copy of the dictionary, where each key appears both in uppercase and lowercase - which is different than what happens when the dictionary is created internally, and also causes it to crash when creating the `SpecialTokensReverse` dictionary (since we have multiple keys pointing to the same value). 
In addition, since the values aren't added to the vocabulary, then they won't be converted to the correct token. 

Example code to reproduce: The same test you pointed me to [here](https://github.com/dotnet/machinelearning/blob/81122c4c48d84fe2f49afd430a2c8fc214311baa/test/Microsoft.ML.Tokenizers.Tests/BertTokenizerTests.cs#L60), with the following change:

Replace the line that creates the BertTokenizer with this:

```cs
var specialTokens = new Dictionary<string, int>() {
    { "[PAD]", 0 },
    { "[UNK]", 1 },
    { "[CLS]", 2 },
    { "[SEP]", 3 },
    { "[MASK]", 4 },
};
// Create two separate options, since during the create the dictionary is manipulated and the Options variables are mutable. 
var options1 = new BertOptions() { SpecialTokens = specialTokens.ToDictionary() };
var options2 = new BertOptions() { SpecialTokens = specialTokens.ToDictionary() }; 
BertTokenizer[] bertTokenizers = [BertTokenizer.Create(vocabFile, options1), BertTokenizer.Create(vocabStream, options2)];
```

This will throw an error during the assert stage, since the first token will be converted to an `[UNK]` since a lowercase `[cls]` doesn't exist in the vocabulary. 

In order for this to work, you need to take the change from my commit in this PR, and also replace the following line:

https://github.com/dotnet/machinelearning/blob/81122c4c48d84fe2f49afd430a2c8fc214311baa/src/Microsoft.ML.Tokenizers/Model/WordPieceTokenizer.cs#L50

with:
```cs
SpecialTokensReverse = options.SpecialTokens is not null ? options.SpecialTokens.GroupBy(kvp => kvp.Value).ToDictionary(g => g.Key, g => g.First().Key) : null;
```

With all this being said, the simplest solution is to make the special tokens handling be the same in both cases - create a copy of the dictionary with only the lowercase keys, and add them explicitly to the vocabulary (as is done in AddSpecialToken).

What do you think?

tarekgh commented:
Thanks @shaltielshmid!

> 1] There is a hidden assumption here that the SpecialTokens dictionary keys will be lowercased.

That is right when the lower-case option is specified. This is important because we lower case the input text we tokenize before processing it. 

> 2] What if the lowercased special token already existed in the vocabulary, as a regular token? Of course, this is not a very likely scenario, but worth considering.

This doesn't matter much. the current code is just doing `vocab[new StringSpanOrdinalKey(normalizedToken)] = id;`. So, it will just override the id value there which even correct the problem if the vocab was using wrong id for that special token and specifying lowercasing. 

> 3] In the WordPieceTokenizer.cs file, a reversed dictionary is created (SpecialTokensReverse) - does it make sense that this dictionary value points to the lowercased token?

I don't think the reverse special tokens need to include the lowercased tokens. The intention of this is always to return the normalized form of the special token. 

> It seems like the usage of this dictionary is just to check if an ID is a special token, so to solve this we can just replace this variable with a HashSet of the IDs.

Although we have this as internal property today, it is possible we'll need to expose it in the future to allow mapping the special token id to the string token. Having it as a dictionary should be correct to use for that.

I stand corrected regarding adding the lowered cased special tokens to the vocab. I forgot we already doing that when the special tokens not provided in the options. Yes, we need to have consistent behavior whether the special tokens are provided or not.
We could call `AddSpecialToken` on the provided special tokens

```C#
                if (options.SpecialTokens is not null)
                {
                    if (lowerCase)
                    {
                        Dictionary<string, int> dic = options.SpecialTokens.ToDictionary(kvp => kvp.Key, kvp => kvp.Value);

                        foreach (var kvp in options.SpecialTokens)
                        {
                            AddSpecialToken(vocab, dic, kvp.Key, lowerCase: true);
                        }
                        
                        // I commented the following line too to avoid overwriting the special tokens in the options.
                        // options.SpecialTokens = dic; 
                    }
                }
```

One thing we can consider is not overwriting the special tokens inside the options. This can be done by storing the special tokens in a local variable and use it in the line 

```C#
options.PreTokenizer ??= options.ApplyBasicTokenization ? PreTokenizer.CreateWordOrPunctuation(options.SplitOnSpecialTokens ? options.SpecialTokens : null) : PreTokenizer.CreateWhiteSpace();
```

This should help not needing the change 

```C#
SpecialTokensReverse = options.SpecialTokens is not null ? options.SpecialTokens.GroupBy(kvp => kvp.Value).ToDictionary(g => g.Key, g => g.First().Key) : null;
```

But I don't mind having this change anyway just in case anyone provides a duplicated Ids special tokens. 

Let me know if there is anything unclear or if you want me to help edit your PR. Thanks for your help!

shaltielshmid commented:
Thank you for the detailed response @tarekgh !

I updated the code as discussed, and added tests for both. 

Two things I'd like to note:

1] When the code dynamically creates the special tokens dictionary, I kept a separate dictionary of the un-normalized tokens which I assigned to the BertOptions.SpecialTokens so that it would propagate onwards to the WordPieceTokenizer class. 
- I kept the tokens not normalized, so that the behavior would match when passing in the tokens explicitly and when the code creates it. 
- We still need to modify the BertOptions variable so that the SpecialTokens gets propagated to the WordPieceTokenizer, even though we would rather not modify it. 

2] This is a general comment about the tokenizer with lowercase - in hugginface's `tokenizers` library, they extract the special tokens *before* the normalization, so that they only match exact-case matches. (To be precise, they store a flag for each special token specifying whether the special token is pre/post normalizing, and then they have two procedures where they extract the special tokens - before normalization and before pre-tokenization). 

This will result in different behavior, where in Python we would have:
```python
from transformers import AutoTokenizer
tok = AutoTokenizer.from_pretrained('bert-base-uncased')
tok.tokenize("[cls] hello")
# Output: ['[', 'cl', '##s', ']', 'hello']
```

And in Microsoft.ML:

```cs
tokenizer = BertTokenizer.Create(vocab_file, options);
tokenizer.EncodeToTokens("[cls] hello", out _)); 
// Output: ["[CLS]", "hello"]
```

Not critical to change, but we should be aware that there is this discrepancy. If you're interested, I'm happy to try and create a separate PR which addresses this issue.

tarekgh commented on src/Microsoft.ML.Tokenizers/Model/BertTokenizer.cs:
```diff
@@ -762,54 +762,68 @@ private static BertTokenizer Create(
+                    else
+                    {
+                        specialTokensDict = options.SpecialTokens?.ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
+                    }
```
We don't have to do it this way allocating more dictionary. what we can do is:

- Define:
IReadOnlyDictionary<string, int>? specialTokensDict = options.SpecialTokens;

- change `specialTokensDict = [];` to soemthing like:

```
Dictionary<string, int> tempSpecialTokens = [];
pecialTokensDict = tempSpecialTokens;
```

- Then pass `tempSpecialTokens` to AddSpecialToken.

You'll need to same idea in the else part when the special tokens not provided in options.

tarekgh commented:
@shaltielshmid your changes look good, I left a comment if you can address it. 

> This will result in different behavior, where in Python we would have:

You need to set `options.SplitOnSpecialTokens = false` when creating the tokenizer. [Python implementation] (https://huggingface.co/transformers/v4.8.2/model_doc/bert.html?highlight=berttokenizer#berttokenizer) uses 
```
never_split (Iterable, optional) – Collection of tokens which will never be split during tokenization. Only has an effect when do_basic_tokenize=True
```

In our implementation we gave better control to special tokens and decided to split them. `options.SplitOnSpecialTokens` is true in our case.

shaltielshmid commented:
Thanks for your comment - great point, I updated the code accordingly.

tarekgh reviewed (approved):
LGTM!

Thanks @shaltielshmid

tarekgh commented:
/backport to release/4.0
